### ISSUE TITLE
depends on docker image??

### ISSUE THREAD

**Issue by BoPeng** [BoPeng/SOS#88]
I am not sure how to let SoS call another `docker_build` step if a `docker_image` is not available. The problem here is that docker does not provide docker images as physical files that SoS can check dependency. 

It is not terribly troublesome because I am running

```
sos run script build
sos run script default
```

to build the docker image before running it, but perhaps we should extend our file-based dependency a little bit.

**Comment by bpeng2000** [comment]
I believe this can be resolved by introducing a few functions that return a file. For example,

```
define docker_image(name):
    #
    if docker has image:
            touch '.sos/_docker_image_hash'
            return '.sos/_docker_image_hash'
```

Then we can do

```
[10]
output: docker_image('something')
docker_build:  tag='something'
   dockerfile

[20]
depends:  docker_image('something')
```

**Comment by bpeng2000** [comment]
This syntax is problematic because we will need to write a signature file when the process is done. It should be something like this but 

```
[10]
output: docker_image('something')
process:
docker_build(dockerfile, tag='something')
docker_image('something').write()

[20]
depends:  docker_image('something')
```

We can probably add a runtime option `signature` that `touch` a specified file.

```
[10]
output: docker_image('something')
process:
docker_build: tag='something', signature=docker_image('something')
  dockerfile

[20]
depends:  docker_image('something')
```

This can actually be extended to create a specified signature file from any text if the action does not create any output.

```
[10]
output: signature('some task')

process:
run: signature=signature('some task')
  no file is generated

[20]
depends:  signature('some task')
```

**Comment by bpeng2000** [comment]
I would propose the following syntax, a function called `signature`, `mark`, or `trait`, that returns an unique ID from simple text or some specific item (e.g. `docker-image`)

```
def trait('type:name'):
    if type == 'image':
         get id of image
    elif no type:
          get md5 of text
    return '.sos/trait' + hex id
```

Define a runtime option `signature`, or `touch`, or `mark` to touch a file (perhaps with some other information) after successful completion of process. We can also extend this to something like `on_succ=do something`.

Then the result of `trait(something)` can be used whenever a file is not actually generated but needed for SoS to function.

**Comment by gaow** [comment]
I'm just wondering if we can get away without introducing another explicit syntax. We hide things like signatures from the users and that feels like SoS magic. Exposing the magic to users makes it less intuitive. What if before actually running commands, we "survey" all the docker files involved and try to create signatures based on their IDs, into a dedicated folder `.sos/.dockers`? I mean if this new function proposed does not have much use beyond dockers.

**Comment by bpeng2000** [comment]
It is for cases when
 
1. output is not a file (e.g. docker image), 
2. there is no output but has to be done before another step (e.g. svn update, git pull, apt-get update, register something), 
3. the output is input (e.g. vtools project) but need to track progress,
4. the output is too complex to write down and would be needed implicitly, namely no explicit filenames are needed (e.g. ???). 

In these cases, users can thrown in a flag file (e.g. `.sos/trait_2fewer23`) from a description  `trait('blah created')` to present the completion of the step.

**Comment by BoPeng** [comment]
Syntax wise we can introduce special filenames such as `docker_image:blah`, `completed:jobname`. SoS can automatically translate these to real filenames such as `.sos/job_hash`. This is not something urgent, let us keep it as an idea here.

**Comment by BoPeng** [comment]
superseded by another ticket.
